### PR TITLE
Ports Metal Detectors from Persistence

### PR DESCRIPTION
**Pull request by XanderDox**
## About The Pull Request

This PR ports metal detectors from Persistence 13. (Originally coded by NataKiller: https://github.com/Persistent-SS13/Persistent-Bay/pull/361/files )

Removes faction-based checks from the metal detectors and replaces them with access checks for security, council, judge and bodyguard (will not go off for these individuals/access)

Plays an alarm sound.

## Why It's Good For The Game

Gives the Police/City Hall Security a quick at a glance method of knowing if someone may be worth a search before allowing them into secure areas. Was also on Cassie's to do lists.

## Changelog
:cl: XanderDox
add: Ported Metal Detectors from Persistence.
soundadd: Ported Alarm Sound from Persistence.

/:cl:

### CODE REVIEW

**Comment by XanderDox** [comment]
One merge conflict resolved, another created. We love the github life.

**Comment by XanderDox** [comment]
@GeneriedJenelle I can't figure out why there is a merge conflict. I've updated stationobjs.dmi with the latest version in the main repo and yet it's still bitching. Plz fix.

File: code/game/machinery/metaldetector.dm [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/obj/machinery/metal_detector
+	name = "metal detector"
+	desc = "A advanced metal detector used to detect weapons."
+	icon = 'icons/obj/stationobjs.dmi'
+	icon_state = "metal_detector"
+	anchored = 1
+	req_one_access = list(access_security, access_judge, access_heads, access_bodyguard)
+	var/list/banned_objects=list(/obj/item/weapon/gun,
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]:
```suggestion
	var/static/list/banned_objects = list(/obj/item/weapon/gun,
```

File: code/game/machinery/metaldetector.dm [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+								/obj/item/weapon/melee,
+								/obj/item/device/transfer_valve,
+								/obj/item/weapon/grenade/,
+								/obj/item/ammo_casing/,
+								/obj/item/ammo_magazine
+								)
+	plane = MOB_PLANE
+	layer = ABOVE_MOB_LAYER
+
+/obj/machinery/metal_detector/New()
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]:
```suggestion
/obj/machinery/metal_detector/Initialize()
```

File: code/game/machinery/metaldetector.dm [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	..()
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]:
```suggestion
	. = ..()
```

File: code/game/machinery/metaldetector.dm [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/metal_detector(src)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(src)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(src)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(src)
+	component_parts += new /obj/item/weapon/stock_parts/console_screen(src)
+	RefreshParts()
+
+/obj/machinery/metal_detector/attackby(obj/item/W, mob/usr)
+/*	if(istype(W, /obj/item/weapon/card/id) || istype(W, /obj/item/device/pda))
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]: Commented out code is bad.

File: code/game/machinery/metaldetector.dm [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	..()
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/metal_detector(src)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(src)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(src)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(src)
+	component_parts += new /obj/item/weapon/stock_parts/console_screen(src)
+	RefreshParts()
+
+/obj/machinery/metal_detector/attackby(obj/item/W, mob/usr)
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]: `usr` should be `user` in this entire proc

Comment (ike709) [reply code/game/machinery/metaldetector.dm]: Also you don't call `..()` at the end of attackby()

File: code/game/machinery/metaldetector.dm [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+			to_chat(usr, "<span class='notice'>\The [src] has been synced to your faction</span>")
+			return*/
+
+	if(default_deconstruction_screwdriver(usr, W))
+		return
+
+	if(default_deconstruction_crowbar(usr, W))
+		return
+
+/obj/machinery/metal_detector/Crossed(var/atom/A)
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]:
```suggestion
/obj/machinery/metal_detector/Crossed(var/mob/living/M)
```

File: code/game/machinery/metaldetector.dm [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	if(istype(A, /mob/living))
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]:
```suggestion
	if(!istype(L))
		return
```

File: code/game/machinery/metaldetector.dm [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		var/mob/living/M = A
+		if(allowed(M))
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]:
```suggestion
	if(allowed(M))
```

File: code/game/machinery/metaldetector.dm [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+			return
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]:
```suggestion
		return
```

File: code/game/machinery/metaldetector.dm [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	var/list/checked_items = list()
+	checked_items += A
+	if(A.contents.len)
+		checked_items += A.contents // Double-recursive check. A gun in a satchel will set it off, a knife in a box in a satchel will too.
+		for(var/obj/O in A.contents) // A knife in a bible in a box in a satchel will not.
+			if(O.contents.len)
+				checked_items += O.contents
+				for(var/obj/OT in O.contents)
+					if(OT.contents.len)
+						checked_items += OT.contents
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]: You should only be looping through the mob's items and the contents of the items on the mob.

File: code/game/machinery/metaldetector.dm [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+	check_items(checked_items)
+	..()
+
+/obj/machinery/metal_detector/proc/check_items(var/list/L)
+	for(var/O in banned_objects)
+		for(var/A in L)
+			if(istype(A, O))
```
Comment (ike709) [inline code/game/machinery/metaldetector.dm]: This is awful.

Comment (XanderDox) [reply code/game/machinery/metaldetector.dm]: It works and it is what was provided to me. Do you have any advice on how to fix it?

File: code/game/machinery/metaldetector.dm (hunk 10 above)
Comment (XanderDox) [reply code/game/machinery/metaldetector.dm]: You assume I know what this means. I only ported this code.

File: code/game/machinery/metaldetector.dm (hunk 5 above)
Comment (XanderDox) [reply code/game/machinery/metaldetector.dm]: @ike709 What is the difference between usr and user if they both work? Also, nothing is calling . . ( ) at the end of attackby?

**Pull request closed without merging**